### PR TITLE
add IPMI polling

### PR DESCRIPTION
**Pull request by VVelox**
Adds polling using ipmi-sensors.

### CODE REVIEW

**Comment by ktsaou** [comment]
Hi, thanks!

Is it possible to call `ipmi_sensors` and `awk` only once at `IPMI_update()`.

As it is, it does 6 `forks()` per second (each external command is a `fork()`).

If you post here the output of `ipmi_sensors` I may be able to help.

**Comment by ktsaou** [comment]
It also does 3x `sudo`, so it is 9 `fork()` per second.

**Comment by VVelox** [comment]
There data needs to be on a separate graph, so not really possible to combine into a single awk statement as AFAIK it would not be possible of manipulating the data separately then for each graph.

ID   | Name          | Type        | Reading    | Units | Event
512  | 12V1          | Voltage     | 11.87      | V     | 'OK'
528  | 12V           | Voltage     | 11.78      | V     | 'OK'
544  | SIO_AVCC      | Voltage     | 3.30       | V     | 'OK'
560  | VCC3.3        | Voltage     | 3.31       | V     | 'OK'
608  | 12V           | Voltage     | 11.78      | V     | 'OK'
736  | 3.3V DUAL     | Voltage     | 3.25       | V     | 'OK'
1792 | P0_VDD        | Voltage     | 1.21       | V     | 'OK'
1808 | P0_VDDNB      | Voltage     | 1.09       | V     | 'OK'
1824 | P1_VDD        | Voltage     | 1.20       | V     | 'OK'
1840 | P1_VDDNB      | Voltage     | 1.09       | V     | 'OK'
1856 | P0_VTT        | Voltage     | 0.75       | V     | 'OK'
1872 | P1_VTT        | Voltage     | 0.74       | V     | 'OK'
1888 | CPU0_VLDT     | Voltage     | 1.21       | V     | 'OK'
1904 | CPU1_VLDT     | Voltage     | 1.19       | V     | 'OK'
1920 | VDDPCIE_RD890 | Voltage     | 1.13       | V     | 'OK'
1936 | VDDC_RD890    | Voltage     | 1.13       | V     | 'OK'
1984 | VDDHTTX_RD890 | Voltage     | 1.20       | V     | 'OK'
2000 | HWM_3VDD      | Voltage     | 3.22       | V     | 'OK'
2016 | HWM_3VSB      | Voltage     | 3.14       | V     | 'OK'
2032 | VBAT          | Voltage     | 3.14       | V     | 'OK'
2048 | DIMM0 Outlet  | Temperature | 46.00      | C     | 'OK'
2064 | DIMM1 Inlet   | Temperature | 33.00      | C     | 'OK'
2080 | SR5670 Case   | Temperature | 75.00      | C     | 'OK'
2096 | PCIE Inlet    | Temperature | 27.00      | C     | 'OK'
2128 | PCIE Outlet   | Temperature | 61.00      | C     | 'OK'
2144 | CPU0_TEMP     | Temperature | 38.00      | C     | 'OK'
2160 | CPU1_TEMP     | Temperature | 35.00      | C     | 'OK'
2272 | REAR_FAN2     | Fan         | 3240.00    | RPM   | 'OK'
2288 | FAN8          | Fan         | N/A        | RPM   | N/A
2304 | CPU0_FAN      | Fan         | 1320.00    | RPM   | 'OK'
2320 | CPU1_FAN      | Fan         | 1200.00    | RPM   | 'OK'
2336 | FRONT_FAN1    | Fan         | 3480.00    | RPM   | 'OK'
2352 | FRONT_FAN2    | Fan         | 3600.00    | RPM   | 'OK'
2368 | FRONT_FAN3    | Fan         | 3480.00    | RPM   | 'OK'
2384 | REAR_FAN1     | Fan         | 3240.00    | RPM   | 'OK'

Also the fork fork and time involved is why I have it set to 5 seconds. 5 or 10 is as low as this check should realistically go.

**Comment by VVelox** [comment]
Ohh!

Hey, just found out that apparently bash will allow multiline variable data.

[kitsune@vixen42 ~]$ a=`sudo ipmi-sensors`
[kitsune@vixen42 ~]$ echo "$a"
ID   | Name          | Type        | Reading    | Units | Event
512  | 12V1          | Voltage     | 11.87      | V     | 'OK'
528  | 12V           | Voltage     | 11.78      | V     | 'OK'
544  | SIO_AVCC      | Voltage     | 3.30       | V     | 'OK'
560  | VCC3.3        | Voltage     | 3.31       | V     | 'OK'
608  | 12V           | Voltage     | 11.78      | V     | 'OK'
736  | 3.3V DUAL     | Voltage     | 3.25       | V     | 'OK'
1792 | P0_VDD        | Voltage     | 1.20       | V     | 'OK'
1808 | P0_VDDNB      | Voltage     | 1.09       | V     | 'OK'
1824 | P1_VDD        | Voltage     | 1.20       | V     | 'OK'
1840 | P1_VDDNB      | Voltage     | 1.09       | V     | 'OK'
1856 | P0_VTT        | Voltage     | 0.74       | V     | 'OK'
1872 | P1_VTT        | Voltage     | 0.74       | V     | 'OK'
1888 | CPU0_VLDT     | Voltage     | 1.20       | V     | 'OK'
1904 | CPU1_VLDT     | Voltage     | 1.19       | V     | 'OK'
1920 | VDDPCIE_RD890 | Voltage     | 1.13       | V     | 'OK'
1936 | VDDC_RD890    | Voltage     | 1.13       | V     | 'OK'
1984 | VDDHTTX_RD890 | Voltage     | 1.20       | V     | 'OK'
2000 | HWM_3VDD      | Voltage     | 3.22       | V     | 'OK'
2016 | HWM_3VSB      | Voltage     | 3.14       | V     | 'OK'
2032 | VBAT          | Voltage     | 3.12       | V     | 'OK'
2048 | DIMM0 Outlet  | Temperature | 46.00      | C     | 'OK'
2064 | DIMM1 Inlet   | Temperature | 33.00      | C     | 'OK'
2080 | SR5670 Case   | Temperature | 76.00      | C     | 'OK'
2096 | PCIE Inlet    | Temperature | 26.00      | C     | 'OK'
2128 | PCIE Outlet   | Temperature | 61.00      | C     | 'OK'
2144 | CPU0_TEMP     | Temperature | 38.00      | C     | 'OK'
2160 | CPU1_TEMP     | Temperature | 35.00      | C     | 'OK'
2272 | REAR_FAN2     | Fan         | 3360.00    | RPM   | 'OK'
2288 | FAN8          | Fan         | N/A        | RPM   | N/A
2304 | CPU0_FAN      | Fan         | 1320.00    | RPM   | 'OK'
2320 | CPU1_FAN      | Fan         | 1200.00    | RPM   | 'OK'
2336 | FRONT_FAN1    | Fan         | 3600.00    | RPM   | 'OK'
2352 | FRONT_FAN2    | Fan         | 3720.00    | RPM   | 'OK'
2368 | FRONT_FAN3    | Fan         | 3600.00    | RPM   | 'OK'
2384 | REAR_FAN1     | Fan         | 3360.00    | RPM   | 'OK'
[kitsune@vixen42 ~]$ 

Will have to modify that. :)

**Comment by VVelox** [comment]
Okay down from 9 to 5.

**Comment by VVelox** [comment]
Wooho! The peaks every 5 seconds are now a lot shorter. :)

As to calling awk less, sadly don't think that can still be avoided.

**Comment by ktsaou** [comment]
`awk` is also magical. Check this: https://github.com/firehol/netdata/blob/b081a001f8181ce34fbd579a9b7b278cdcd4c786/charts.d/nut.chart.sh#L138-L216

So you can do just one `ipmi-sensors` and just one `awk`.

**Comment by ktsaou** [comment]
At the `ipmi-sensors` [man page](https://www.gnu.org/software/freeipmi/manpages/man8/ipmi-sensors.8.html), I see:

>--comma-separated-output
>Output fields in comma separated format.

>--no-header-output
>Do not output column headers. May be useful in scripting.

>--non-abbreviated-units
>Output non-abbreviated units (e.g. 'Amps' instead of 'A'). May aid in disambiguation of units (e.g. 'C' for Celsius or Coulombs).

It may be easier to parse the output with a few of them.

**Comment by ktsaou** [comment]
Also, are you sure `sudo` is needed?
You can become user netdata with this (execute it as root):

```sh
su -s /bin/sh netdata
```

**Comment by VVelox** [comment]
Was just experimenting with awk some more and it appears the variable in it can handle more than one line of text in them.

It's not that the data is hard to parse, but that I can't be certain exactly what all it will return as that will very depending on the server in question. Below is a example output another.

Finding out that awk will not choke the example below ill make it possible.

echo | awk '{a = "\n"; a = "line 1" a a "line 4"; print a}'

[root@vulpes]/home/kitsune# ipmi-sensors
ID  | Name          | Type              | Reading    | Units | Event
4   | System Temp   | Temperature       | 38.00      | C     | 'OK'
71  | CPU Temp      | Temperature       | 34.00      | C     | 'OK'
138 | CPU FAN       | Fan               | 7985.00    | RPM   | 'OK'
205 | SYS FAN       | Fan               | 8355.00    | RPM   | 'OK'
272 | CPU Vcore     | Voltage           | 1.15       | V     | 'OK'
339 | Vichcore      | Voltage           | 1.04       | V     | 'OK'
406 | +3.3VCC       | Voltage           | 3.31       | V     | 'OK'
473 | VDIMM         | Voltage           | 1.52       | V     | 'OK'
540 | +5 V          | Voltage           | 5.06       | V     | 'OK'
607 | +12 V         | Voltage           | 12.19      | V     | 'OK'
674 | +3.3VSB       | Voltage           | 3.26       | V     | 'OK'
741 | VBAT          | Voltage           | 3.15       | V     | 'OK'
808 | Chassis Intru | Physical Security | N/A        | N/A   | 'OK'
875 | PS Status     | Power Supply      | N/A        | N/A   | 'Unrecognized Event = 0100h' 'Unrecognized Event = 0200h' 'Unrecognized Event = 0400h' 'Unrecognized Event = 0800h' 'Unrecognized Event = 1000h' 'Unrecognized Event = 2000h' 'Unrecognized Event = 4000h'
[root@vulpes]/home/kitsune# 

I'll sit down some time over night and knock this out. :)

**Comment by VVelox** [comment]
Sudo is needed as ipmi-sensors needs to be ran as root to work.

[kitsune@vixen42]/home/kitsune% ipmi-sensors
ipmi-sensors: permission denied
Exit 1

Also there is no good way to get to root using su and only allow access to ipmi-sensors.

**Comment by ktsaou** [comment]
ok, run this:

```sh
ipmi-sensors --comma-separated-output | awk -F"," "/,Voltage,[0-9]/ { print \$1 \" = \" \$4 } /,Temperature,[0-9]/ { print \$1 \" = \" \$4 }"
```

I suggest to do this:

1. Decide the type of charts you will support. For example, you can support a chart for voltages, another one for temperatures, a third for currents and a fourth for fans (much like you already do).
2. in `ipmi_create()` use 4 awks to generate the dimensions for each chart. The first column of ipmi-sensors should be the dimension ID (which you will you in `SET` lines) and the name would be the second column of the ipmi-output. Use a multiplier of 100 (netdata parses only integers).
3. in `ipmi_update()` use an awk line, like the one I gave you above, but instead of `print` add each value to an awk [associative array](https://www.chemie.fu-berlin.de/chemnet/use/info/gawk/gawk_12.html), where the index is `$1` and value is `$4`. So, you will have 4 associative arrays, one for each chart.
4. at the `END` section of the awk, print the associative arrays, much like `nut.chart.sh` does.

On a few servers that I tried `ipmi-sensors` it is extremely slow (a few seconds per run). This will slow down all other `chart.d.plugin` modules.  I have a solution though, that will allow us to avoid `sudo` too. But I will show you this, once we have all above in place.

**Comment by LeeNX** [comment]
I think there must be some mind readers in this repo ... Was just starting to play with IPMI stuff. Here is some of my input.  
  
Other commands that could be used are ```ipmitool sensor``` and ```ipmimonitoring```, but give must the same and seem a little slower. I think that is because ```ipmi-sensors``` caches the sensor data repository (SDR), but this could also just be on my system.  
  
My output ...  
```
clintontaylor@spartan:~$ sudo ipmi-sensors
ID | Name             | Type                     | Reading    | Units | Event
1  | Pwr Unit Status  | Power Unit               | N/A        | N/A   | 'OK'
2  | IPMI Watchdog    | Watchdog 2               | N/A        | N/A   | 'OK'
3  | Physical Scrty   | Physical Security        | N/A        | N/A   | 'OK'
4  | SMI TimeOut      | OEM Reserved             | N/A        | N/A   | 'OK'
5  | System Event Log | Event Logging Disabled   | N/A        | N/A   | 'OK'
6  | System Event     | System Event             | N/A        | N/A   | 'OK'
7  | Button           | Button/Switch            | N/A        | N/A   | 'OK'
8  | PCH Therm Trip   | Temperature              | N/A        | N/A   | 'OK'
9  | BMC Board TEMP   | Temperature              | 31.00      | C     | 'OK'
10 | Board Inlet TEMP | Temperature              | 23.00      | C     | 'OK'
11 | Processor Fan    | Fan                      | 980.00     | RPM   | 'OK'
12 | System Fan 4     | Fan                      | 2058.00    | RPM   | 'OK'
13 | Sys Fan 1        | Fan                      | 1176.00    | RPM   | 'OK'
14 | PS1 Status       | Power Supply             | N/A        | N/A   | 'OK'
15 | PS1 Power In     | Other Units Based Sensor | N/A        | W     | N/A
16 | PS1 Temperature  | Temperature              | N/A        | C     | N/A
17 | P1 Status        | Processor                | N/A        | N/A   | 'Processor Presence detected'
18 | P1 Therm Margin  | Temperature              | -78.00     | C     | 'OK'
19 | P1 Therm Ctrl %  | Temperature              | 0.00       | %     | 'OK'
20 | P1 ERR2 TimeOut  | Processor                | N/A        | N/A   | 'OK'
21 | CATERR           | Processor                | N/A        | N/A   | 'OK'
22 | MSID Misatch     | Processor                | N/A        | N/A   | 'OK'
23 | CPU Missing      | Processor                | N/A        | N/A   | 'OK'
24 | P1 VRD Hot       | Temperature              | N/A        | N/A   | 'OK'
25 | Mem P1 Thrm Mrgn | Temperature              | -54.00     | C     | 'OK'
26 | Mem P1 Thrm Trip | Memory                   | N/A        | N/A   | 'OK'
27 | BB +12.0V        | Voltage                  | 11.88      | V     | 'OK'
28 | BB +5.0V         | Voltage                  | 5.02       | V     | 'OK'
29 | BB +3.3V         | Voltage                  | 3.24       | V     | 'OK'
30 | BB +5.0V STBY    | Voltage                  | 4.92       | V     | 'OK'
31 | BB +3.3V AUX     | Voltage                  | 3.25       | V     | 'OK'
32 | BB P1 Vcc        | Voltage                  | 1.08       | V     | 'OK'
33 | BB +1.5V P1 MEM  | Voltage                  | 1.49       | V     | 'OK'
34 | BB +3V Vbat      | Voltage                  | 3.22       | V     | 'OK'
35 | BB P1 Vccp       | Voltage                  | 1.03       | V     | 'OK'
36 | BB P1 VccUSA     | Voltage                  | 0.91       | V     | 'OK'
37 | BB +1.05V PCH    | Voltage                  | 1.03       | V     | 'OK'
38 | BB +1.05V AUX    | Voltage                  | 1.03       | V     | 'OK'
39 | BB +12.0V V1     | Voltage                  | 11.88      | V     | 'OK'
40 | BB +1.5V AUX     | Voltage                  | 1.47       | V     | 'OK'
44 | NM Capabilities  | OEM Reserved             | N/A        | N/A   | N/A
```
  
Not sure if more ideas for IPMI monitoring is best stated here or in another issue?

**Comment by VVelox** [comment]
Any sort of array stuff is completely unnecessary as it turns out it will handle \n in variable data with out issue.

Given this it is possible of just using BEGIN statement to put together beginning of each, process each line and add it to the proper variable, and then use END to pull the three strings together.

**Comment by ktsaou** [comment]
freeipmi plugin written in C is in PR #1881

**Comment by ktsaou** [comment]
merged it.
